File: src/lib/api/repo.ts
import {
  array,
  literal,
  number,
  object,
  optional,
  string,
  union,
  z,
} from "zod";

import { authorSchema } from "./author";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const repoSchema = object({
  rid: string(),
  payloads: object({
    "xyz.radicle.project": object({
      data: object({
        name: string(),
        description: string(),
        defaultBranch: string(),
      }),
      meta: object({
        head: string(),
        patches: object({
          open: number(),
          draft: number(),
          archived: number(),
          merged: number(),
        }),
        issues: object({
          open: number(),
          closed: number(),
        }),
      }),
    }),
  }),
  delegates: array(authorSchema),
  threshold: number(),
  visibility: union([
    object({ type: literal("public") }),
    object({ type: literal("private"), allow: optional(array(string())) }),
  ]),
  seeding: number(),
});

export type Repo = z.infer<typeof repoSchema>;

File: src-tauri/src/types/repo.rs
use serde::Serialize;
use serde_json::Value;
use ts_rs::TS;

use radicle::identity::RepoId;

use super::cobs::Author;

/// Repos info.
#[derive(Serialize, TS)]
#[ts(export)]
pub struct RepoInfo {
    pub payloads: SupportedPayloads,
    #[ts(as = "Vec<Author>")]
    pub delegates: Vec<Author>,
    pub threshold: usize,
    #[ts(type = "{ type: 'public' } | { type: 'private', allow?: string[] }")]
    pub visibility: radicle::identity::Visibility,
    #[ts(as = "String")]
    pub rid: RepoId,
    pub seeding: usize,
}

#[derive(Serialize, TS)]
#[ts(export)]
pub struct SupportedPayloads {
    #[serde(rename = "xyz.radicle.project")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    #[ts(type = r#"{
  data: {
    defaultBranch: string,
    description: string,
    name: string,
  },
  meta: {
    head: string,
    issues: {
      open: number,
      closed: number,
    },
    patches: {
      open: number,
      draft: number,
      archived: number,
      merged: number,
    }
    lastCommit: number,
  }
}"#)]
    pub project: Option<Value>,
}
